internal/stacks/stackruntime/internal/stackeval/component_config.go

```diff
@@ -336,6 +338,10 @@ func (c *ComponentConfig) CheckProviders(ctx context.Context, phase EvalPhase) (
 				})
 				continue
 			}
+		} else if actualTy.Equals(cty.DynamicPseudoType) {
```

Comment:
This is a bit of a tricky edge in `cty` that gets us into trouble sometimes. I'm not sure if this is an example of that trouble, but I'm going to try to describe the trouble and hopefully you can decide if it matters. :grinning: 

The `null` keyword in HCL represents what `cty` would call `cty.NullVal(cty.DynamicPseudotType)` -- a null value of an unknown type. That would cause this expression to return `true`, even though I _think_ we're not intending to accept `null` as a provider configuration reference here:

```hcl
  providers = {
    aws = null
  }
```

To deal with that we've typically inserted an extra check like `if result.Value.IsNull()` before doing any type checking, so that we can deal with that error case first and just treat nulls of any type as equally invalid.

All of that said then: I don't _think_ it makes sense to allow `null` here because any provider configuration slot a module declares is always required. If that's true then I'd suggest adding an extra check up at the top like I described above. If there _is_ some reason to allow null then this might be okay, though to make that case more visible in the code I might still suggest handling it as a separate if condition since it's a common mistake to forget about the possibility of nulls when maintaining existing code.

Reply:
(I notice below there's a comment saying that `component_instance.go` is responsible for checking whether the value is `null`, so maybe just type checking here is sufficient after all. I suggest checking that comment is actually true, though!)

Reply:
Yeah I think I'm on the entirely wrong track here, I think your assessment that this is `cty.NullVal(cty.DynamicPseudoType)` or sth similar is right. I looked into the reference extraction and found that in the `EvalScope` the provider values are from my point of view correct, it's a map of provider configurations. I couldn't find where the property access is handled / should be handled, maybe this issue is too in the weeds for me right now 🙈